# **Assigment no. 3**




**Title:** Assignment on Improving Performance of Classifier Models   
**Aim:** Implement Naive Bayes Classifier on SMSSpam Filtering system 

**Objectives:**

A. Apply Data pre-processing (Label Encoding, Data Transformation….) 

B. Perform data-preparation (Train-Test Split)

C. Apply at least two Machine Learning Algorithms and Evaluate Models

D. Apply Cross-Validation and Evaluate Models and compare performance.

E. Apply Hyper parameter tuning and evaluate models and compare performance.




In [3]:
# Importing the necessary libraries
import pandas as pd
import numpy as np

In [5]:
sms_data = pd.read_csv(r'C:\Users\rushikesh\Downloads\SMSSpamCollection.txt', delimiter = "\t", header=None)

In [7]:
sms_data.head()

0                                                  1
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...

We import ‘re’ package and remove punctuation, special characters and convert all characters to lower case.

Now, we need to split a message into words to remove stop-words and to perform stemming.

We import English stop-words from the NLTK package and removed them if found in the sentence.

While removing stop-words, we perform stemming that is if the word is not a stop-word, it will be converted to its root form. This is called stemming.

Then we are joining the words in the list again to form a message without any stop-words and all words will be present in its
root form. Then we append messages to the corpus.

In [9]:
# import packages
import re
import nltk
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer


### Creating a corpus of refined messages: 

In [11]:
for i in range (0,len(sms_data)):
    print(sms_data)


         0                                                  1
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...
...    ...                                                ...
5567  spam  This is the 2nd time we have tried 2 contact u...
5568   ham               Will ü b going to esplanade fr home?
5569   ham  Pity, * was in mood for that. So...any other s...
5570   ham  The guy did some bitching but I acted like i'd...
5571   ham                         Rofl. Its true to its name

[5572 rows x 2 columns]
         0                                                  1
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to wi

IOPub data rate exceeded.
The Jupyter server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--ServerApp.iopub_data_rate_limit`.

Current values:
ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
ServerApp.rate_limit_window=3.0 (secs)



In [15]:
!pip install nltk


In [17]:
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer
stemming = PorterStemmer()
corpus = []

for i in range (0,len(sms_data)):
    s1 = re.sub('[^a-zA-Z]',repl = ' ',string = sms_data[1][i])
    s1.lower()
    s1 = s1.split()
    s1 = [stemming.stem(word) for word in s1 if word not in set(stopwords.words('english'))]
    s1 = ' '.join(s1)
    corpus.append(s1)

corpus[50]

'what think first time saw class'

In [18]:
# We’ll import Count-Vectorizer from Scikit-learn library to tokenize and vectorize the text
from sklearn.feature_extraction.text import CountVectorizer
countvectorizer =CountVectorizer()

By using this Count-Vectorizer we’ll tokenize a collection of text documents and built a vocabulary, this vocabulary is also used to encode new documents.

To use this Count-Vectorizer, first, we’ll create an instance of Count-Vectorizer class.

Then fit () function is used to learn vocabulary from the document and the transform() function is used to encode as a vector.

In [25]:
x = countvectorizer.fit_transform(corpus).toarray()

# Now we have a numeric vector that has been converted from a string of text.
print(x)

[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]


In [27]:
y = sms_data[0].values
print(y)

['ham' 'ham' 'spam' ... 'ham' 'ham' 'ham']


In [29]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size = 0.3, stratify=y,random_state=100)

# 1. Multinomial Naïve Bayes.

In [47]:
from sklearn.naive_bayes import MultinomialNB
multinomialnb = MultinomialNB()
multinomialnb.fit(x_train,y_train)

MultinomialNB()

In [43]:
y_pred = multinomialnb.predict(x_test)

In [49]:
y_pred

array(['ham', 'ham', 'ham', ..., 'spam', 'ham', 'ham'], dtype='<U4')

# 2. SVM

In [ ]:
from sklearn.svm import LinearSVC
linearsvc = LinearSVC()
linearsvc.fit(x_train,y_train)

In [ ]:
y_pred = linearsvc.predict(x_test)

In [39]:
y_pred

array(['ham', 'ham', 'ham', ..., 'spam', 'ham', 'ham'], dtype='<U4')

# Results of our Models

###  Multinomial Naïve Bayes.

In [37]:
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.metrics import accuracy_score

print(classification_report(y_test,y_pred))
accuracy_score(y_test,y_pred)

              precision    recall  f1-score   support

         ham       0.99      0.98      0.99      1448
        spam       0.89      0.94      0.92       224

    accuracy                           0.98      1672
   macro avg       0.94      0.96      0.95      1672
weighted avg       0.98      0.98      0.98      1672



0.9766746411483254

# **Conclusion:**

Two classifiers Naive Bayes and SVM is implemented on SMSSpam dataset. 
the dataset is in textual format. Countvectorizer is used to convert these text data into vector format.And, the accuracy achieved by Naive Bayes classifier is 98%.